File: src/CsRopExample/MessageLoggingHandler.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using System.IO;

namespace CsRopExample
{
    // /// <summary>
    // /// Logging code 
    // /// </summary>
    // public class MessageLoggingHandler : MessageProcessingHandler
    // {

    //     protected override HttpRequestMessage ProcessRequest(HttpRequestMessage request, CancellationToken cancellationToken)
    //     {
    //         var correlationId = string.Format("{0}{1}", DateTime.Now.Ticks, Thread.CurrentThread.ManagedThreadId);
    //         var requestInfo = string.Format("{0} {1}", request.Method, request.RequestUri);
    //         var message = request.Content.ReadAsStringAsync().Result;
    //         Debug.WriteLine("[HTTP]Request: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message);
    //         return request;
    //     }

    //     protected override HttpResponseMessage ProcessResponse(HttpResponseMessage response, CancellationToken cancellationToken)
    //     {
    //         var correlationId = string.Format("{0}{1}", DateTime.Now.Ticks, Thread.CurrentThread.ManagedThreadId);
    //         var requestInfo = string.Format("{0} {1}", response.RequestMessage.Method, response.RequestMessage.RequestUri);
    //         var message = response.Content != null 
    //             ? response.Content.ReadAsStringAsync().Result 
    //             : "[no body]";
    //         Debug.WriteLine("[HTTP]Response: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message);
    //         return response;
    //     }
    // }
    // https://exceptionnotfound.net/using-middleware-to-log-requests-and-responses-in-asp-net-core/
    // https://devblogs.microsoft.com/aspnet/re-reading-asp-net-core-request-bodies-with-enablebuffering/
    public class LogginMiddleware
    {
        private readonly RequestDelegate _next;

        public LogginMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        private async Task<string> FormatResponse(HttpResponse response)
        {
            if (!response.Body.CanRead || !response.Body.CanSeek)
            {
                return "Resonpse cannot be read";
            }
            //We need to read the response stream from the beginning...
            response.Body.Seek(0, SeekOrigin.Begin);

            //...and copy it into a string
            var sr = new StreamReader(response.Body);
            {
                string text = await sr.ReadToEndAsync();

                //We need to reset the reader for the response so that the client can read it.
                response.Body.Seek(0, SeekOrigin.Begin);

                //Return the string for the response, including the status code (e.g. 200, 404, 401, etc.)
                return $"{response.StatusCode}: {text}";
            }
            
        }

        public async Task Invoke(HttpContext context)
        {
            var request = context.Request;
            request.EnableBuffering();
            var correlationId = string.Format("{0}{1}", DateTime.Now.Ticks, Thread.CurrentThread.ManagedThreadId);
            var requestInfo = string.Format("{0} {1}", request.Method, request.Path);

            
            {
                //We now need to read the request stream.  First, we create a new byte[] with the same length as the request stream...
                var buffer = new byte[Convert.ToInt32(request.ContentLength)];

                //...Then we copy the entire request stream into the new buffer.
                await request.Body.ReadAsync(buffer, 0, buffer.Length);

                //We convert the byte[] into a string using UTF8 encoding...
                var message = System.Text.Encoding.UTF8.GetString(buffer);
                // Reset the request body stream position so the next middleware can read it
                request.Body.Position = 0;
                
                Console.WriteLine("[HTTP]Request: {0}, {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message);
            }
            //Copy a pointer to the original response body stream
            var originalBodyStream = context.Response.Body;
            //Create a new memory stream...
            using (var responseBody = new MemoryStream())
            {
                //...and use that for the temporary response body
                context.Response.Body = responseBody;

                //Continue down the Middleware pipeline, eventually returning to this class
                await _next(context);

                //Format the response from the server
                var response = context.Response;

                //We need to read the response stream from the beginning...
                response.Body.Seek(0, SeekOrigin.Begin);

                //...and copy it into a string
                using(var sr = new StreamReader(response.Body))
                {
                    string text = await sr.ReadToEndAsync();

                    //We need to reset the reader for the response so that the client can read it.
                    response.Body.Seek(0, SeekOrigin.Begin);

                    //Return the string for the response, including the status code (e.g. 200, 404, 401, etc.)
                    var message = $"{response.StatusCode}: {text}";
                    //TODO: Save log to chosen datastore
                    Console.WriteLine("[HTTP]Response: {0}, {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message);

                    //Copy the contents of the new memory stream (which contains the response) to the original stream, which is then returned to the client.
                    await responseBody.CopyToAsync(originalBodyStream);
                }
                
            }
        }
    }
}

File: src/FsRopExample/MessageLoggingHandler.fs
namespace FsRopExample

//open System
//open System.Diagnostics
//open System.Net.Http
//open System.Threading

///// Logging code 
//type MessageLoggingHandler() =
    //inherit MessageProcessingHandler()

    //override this.ProcessRequest(request:HttpRequestMessage , _:CancellationToken) =
    //    let correlationId = sprintf "%i%i" DateTime.Now.Ticks Thread.CurrentThread.ManagedThreadId
    //    let requestInfo = sprintf "%O %O" request.Method  request.RequestUri
    //    let message = request.Content.ReadAsStringAsync().Result
    //    Debug.WriteLine("[HTTP]Request: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message)
    //    request

    //override this.ProcessResponse(response:HttpResponseMessage, _:CancellationToken) =
        //let correlationId = sprintf "%i%i" DateTime.Now.Ticks Thread.CurrentThread.ManagedThreadId
        //let requestInfo = sprintf "%O %O" response.RequestMessage.Method  response.RequestMessage.RequestUri
        //let message = if response.Content <> null then response.Content.ReadAsStringAsync().Result else "[no body]"
        //Debug.WriteLine("[HTTP]Response: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message)
        //response

open System
open System.Threading
open Microsoft.AspNetCore.Http
open System.IO

type LoggingMiddleware(next:RequestDelegate) = 
    member this.Invoke (context: HttpContext) = 
        context.Request.EnableBuffering()
        let correlationId = sprintf "%i%i" DateTime.Now.Ticks Thread.CurrentThread.ManagedThreadId
        let requestInfo = sprintf "%O %O" context.Request.Method  context.Request.Path
        let len = 
            if context.Request.ContentLength.HasValue then (int context.Request.ContentLength.Value)
            else 0

        let buffer: byte array = Array.zeroCreate len

        context.Request.Body.Read(buffer, 0, buffer.Length) |> ignore

        let message = System.Text.Encoding.UTF8.GetString(buffer)

        context.Request.Body.Seek(0L, SeekOrigin.Begin) |> ignore

        System.Console.WriteLine("[HTTP]Request: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message)

        let originalBodyStream = context.Response.Body

        use responseBody = new MemoryStream ()
        context.Response.Body <- responseBody

        let task = next.Invoke(context)

        context.Response.Body.Seek(0L, SeekOrigin.Begin)|> ignore

        use sr = new StreamReader(context.Response.Body)
        let text = sr.ReadToEnd()
        context.Response.Body.Seek(0L, SeekOrigin.Begin) |> ignore
        let msg = sprintf "%i %O" context.Response.StatusCode  text

        System.Console.WriteLine("[HTTP]Response: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, msg)

        responseBody.CopyTo(originalBodyStream)

        task